### PR TITLE
[UnifyLoopExits] Never generate phis of only `undef` values

### PR DESCRIPTION
**Pull request by jdoerfert**

### CODE REVIEW

File: llvm/lib/Transforms/Utils/BasicBlockUtils.cpp [hunk 1]
```diff
@@ -1909,22 +1909,29 @@ static void reconnectPhis(BasicBlock *Out, BasicBlock *GuardBlock,
...
         V = Phi->removeIncomingValue(In, false);
+        AllUndef &= isa<UndefValue>(V);
       }
       NewPhi->addIncoming(V, In);
     }
     assert(NewPhi->getNumIncomingValues() == Incoming.size());
+    Value *NewV = NewPhi;
+    if (AllUndef) {
+      NewPhi->eraseFromParent();
+      NewV = UndefValue::get(Phi->getType());
```
Comment (arsenm) [inline llvm/lib/Transforms/Utils/BasicBlockUtils.cpp:1927]: Should this try to preserve poison?

Comment (jdoerfert) [reply llvm/lib/Transforms/Utils/BasicBlockUtils.cpp:1927]: As of right now, this emits undef by default. We likely want to change both UndefValue::get to poison. Should I include that and track undef/poison separately?

Comment (arsenm) [reply llvm/lib/Transforms/Utils/BasicBlockUtils.cpp:1927]: It probably makes little difference, can switch these to poison as a second step?

Comment (jdoerfert) [reply llvm/lib/Transforms/Utils/BasicBlockUtils.cpp:1927]: OK, let's do this first, then we switch to poison.

Comment (ssahasra) [reply llvm/lib/Transforms/Utils/BasicBlockUtils.cpp:1927]: I didn't really understand. Are we agreeing that it's okay to temporarily ignore `UndefValue` that are actually poison in this patch?

Comment (arsenm) [reply llvm/lib/Transforms/Utils/BasicBlockUtils.cpp:1927]: All poisons are undef. Not all undefs are poison. It's safe, though suboptimal, to replace poison with undef

**Review by arsenm — APPROVED** [review]

**Comment by RSilicon** [comment]
Does this affect codegen correctness or is this an enhancement? If it is the former, should this be backported to 19.x?

**Comment by jdoerfert** [comment]
> Does this affect codegen correctness or is this an enhancement? If it is the former, should this be backported to 19.x?

Didn't see this. It's just an enhancement. No functional change.

**Pull request merged**